### ISSUE TITLE
Jobs stuck in "inactive" state

### ISSUE THREAD

**Issue by elbuo8** [Automattic/kue#660]
We have been getting `Redis connection gone from end event` from our `consumers`. I've been investigating what might be causing this and came to notice that whenever our provider runs a backup, around 20-30 mins after, this error came up (hasn't happened in a bit).

If the error does come up, the consumer marks the job as `inactive` and doesn't process it (or so it seems). When restarted, the `watchdog` picks the `inactive` jobs in the first place and everything runs smoothly. 

Is there a way to mitigate the connection error better?

**Comment by behrad** [comment]
I haven't seen this error that much and don't know about its details, however seems that when it happens there's no connectivity with redis, if so kue can't do too much with existing connection. So we should fail fast or reconnect in the middle of the current operation and continue, which the later should investigated if it can implemented side-effect-free.

**Comment by elbuo8** [comment]
How does it become `inactive` if the connection is not available at all?

**Comment by behrad** [comment]
I can't tell you until i know whats happening on your side
first tell me which version of kue/redis are you on!? 
it seems only one of kue redis connections are getting `connection gone` and others are working.

Try to capture whats happening when your client backs up. Is redis under pressure? is machine taking to much CPU then? Is your network bandwidth near full then?

**Comment by elbuo8** [comment]
Kue version is `0.9.3`.
Redis version is `2.8.21`

Everything is idle. Staging env with basically no usage.

**Comment by nbhattarai** [comment]
I was also having similar issues where some jobs were just idle. I added `queue.watchStuckJobs()` which basically is fixing those idle jobs automatically.

**Comment by elbuo8** [comment]
We do use that and it works great whenever the instance is restarted.

**Comment by elbuo8** [comment]
We do use that and it works great whenever the instance is restarted.

**Comment by Grandiz** [comment]
same to me. i have a lot of inactive jobs. redis connection is ok.
queue.watchStuckJobs() help me, but how offen should i use this? maybe someone have better way to fix it?

**Comment by behrad** [comment]
for now you can simply call `queue.watchStuckJobs()` when your node process starts.

**Comment by Grandiz** [comment]
begrad, but what i start nodejs process only once?
i read your msg here https://github.com/Automattic/kue/issues/670 , but still don't get how to watch inactive jobs. i wonder if you could give some workaround

**Comment by behrad** [comment]
thats something else talked about in that topic @Grandiz 
Here you have a problem where some jobs are **STUCK** in inactive state being orphan (without a chance to get processed when a worker is idle) in this case, calling `watchStuckJobs` at the start of your process with check for them trough out your process life.
However that topic is for the case where you want capture size of your inactive jobs, to implement for example a kind of back pressure or so... :)

**Comment by elbuo8** [comment]
Would it be sane to invoke watchStuckJobs every X amount of minutes?

**Comment by behrad** [comment]
no, `watchStuckJobs` has an internal setInterval to check inactive jobs for you, for why it is actually named `watch` stuck jobs

**Comment by elbuo8** [comment]
Then I'm unsure of what is happening. Our servers are under no stress at all. The watcher is invoked at the beggining in all the threads of the worker yet, jobs still end up stucked.

**Comment by behrad** [comment]
stucked in inactive? 
will they be get processed after you restart kue? (without calling watchStuckJobs)

If yes, then it's your workers problem getting stuck...

**Comment by tsturzl** [comment]
Check your logs for crashes. If an API instance that started a job crashes, I have the issue where inactive jobs remain in inactive state even if watchStuckJobs is invoked.

**Comment by pcothenet** [comment]
The symptoms described by @elbuo8 looks a lot like the ones I'm having.

My jobs start processing correctly, then after a while, the new ones just go to inactive. There's no signs of failed job (and I've spent the last few days making sure of that). If I restart the consumer, things pick up just fine (then stop again after a while).

This happen on three different workers (in three different places_ consuming from the same redis, which lets me think the problem is around kue.

**Comment by behrad** [comment]
https://github.com/Automattic/kue/issues/660#issuecomment-123643724

**Comment by pcothenet** [comment]
FWIW, I changed my redis hosting (from RedisLabs to Azure) and that completely solved the issue. So I'm not convinced this was entirely because of the worker. Are you sure this couldn't be something else?

**Comment by behrad** [comment]
most of problems are two fold
1) unhandled errors in worker's process function
2) unstable/poor redis connections

**Comment by fjorgemota** [comment]
Hello.

I suffered from this problem yesterday, when I implemented a type of "mock" to use ioredis instead of node_redis (based on changes from the pull request co-related to that), implementing an EventEmitter with interface similar to what kue use in node_redis, but using ioredis.

Well, the mock initially suffered from some problems, like methods that kue use and are missing in the object, and that caused exactly this problem, from jobs stucking in "inactive" state. Initially, I through that was a bug in my fake object, but, when I changed to node_redis client, the problem yet occurred. Even when calling `queue.watchStuckJobs()` in the start of the program.

However, I discovered something at least interesting: If I change the priority of the job via web UI (even to a lower priority), kue automatically detects the change and runs the job automatically. This gives me the idea of creating a timer that detects tasks too older and change it's priority, but I am not getting the problem again here to test, or maybe just checking this in the start of the program, maybe.

I will try to remove some methods from my fake client object to test if I can cause the problem again, considering that the problems can occur just in some methods of the redis client. For now, I suggest trying that solution: Try to change the priority of a job that is stuck in "inactive" state and see if the job starts again.

PS: Sorry for my poor english, btw.

**Comment by emaddoma** [comment]
@pcothenet We had same issues on RedisCloud.  Switched to Heroku Redis and it went away

**Comment by dinhanhhuy** [comment]
Any update, i follow the reference below but don't see any update to fix this issue :'(

**Comment by lucas-d-nguyen-08** [comment]
Hi @behrad , Does queue.watchStuckJobs() can fix my issue? :(
All of my job are still in queue
![image](https://user-images.githubusercontent.com/111754907/216753524-6362b4e4-576e-4579-9438-42ed39f81f3b.png)
